### PR TITLE
Calculate version range in init-compiler.sh script

### PR DESCRIPTION
**Pull request by am11**
Guesstimations based on https://releases.llvm.org/ and https://gcc.gnu.org/releases.html, which is close enough.
Testing CI, shall upstream once green.

### CODE REVIEW

File: eng/common/native/init-compiler.sh [hunk 1]
```diff
@@ -97,25 +101,13 @@ if [ -z "$CLR_CC" ]; then
 
         if [ -z "$majorVersion" ]; then
             if ! command -v "$compiler" > /dev/null; then
-                echo "Error: No usable version of $compiler found."
+                echo "Error: No compatible version of the $compiler was found within the range of $minVersion to $maxVersion. Upgrade your toolchain or set CLR_CC and CLR_CXX explicitly."
```
Comment (am11) [inline eng/common/native/init-compiler.sh:104]:
```suggestion
                echo "Error: No compatible version of the $compiler was found within the range of $minVersion to $maxVersion. Please upgrade your toolchain or specify the compiler explicitly using CLR_CC and CLR_CXX environment variables."
```

**Comment by akoeplinger** [comment]
Clever, but one downside is that we'd automatically opt-in to newer versions even in servicing branches.

**Comment by am11** [comment]
> one downside is that we'd automatically opt-in to newer versions even in servicing branches.

When someone will update the container reference (all linux legs are using containers these days),  at that time it will show up the compiler error. This is same as what we do today. The only difference is we don't have to update this file in addition to backporting the compilation fixes.

Containers only have one toolchain version installed at a time.

**Comment by am11** [comment]
> one downside is that we'd automatically opt-in to newer versions even in servicing branches.

When someone will update the container reference in servicing branch (all linux legs are using containers these days),  at that time it will show up red with the compiler error. This is same as what we do today. The only difference with this change is that we won't need to update this file in addition to backporting the compilation fixes.

Containers only have one toolchain version installed at a time.

File: eng/common/native/init-compiler.sh [hunk 2]
```diff
@@ -84,43 +82,28 @@ if [ -z "$CLR_CC" ]; then
...
-        if [ "$compiler" = "clang" ]; then versions="18 17 16 15 14 13 12 11 10 9 8 7 6.0 5.0 4.0 3.9 3.8 3.7 3.6 3.5"
-        elif [ "$compiler" = "gcc" ]; then versions="14 13 12 11 10 9 8 7 6 5 4.9"; fi
-
-        for version in $versions; do
-            _major="${version%%.*}"
-            [ -z "${version##*.*}" ] && _minor="${version#*.}"
-            desired_version="$(check_version_exists "$_major" "$_minor")"
-            if [ "$desired_version" != "-1" ]; then majorVersion="$_major"; break; fi
+        i="$((maxVersion + 1))" # +1 for headspace
+        minVersion=8
```
Comment (am11) [inline eng/common/native/init-compiler.sh:86]:
```suggestion
        maxVersion="$((maxVersion + 1))" # +1 for headspace
        i="$maxVersion"
        minVersion=8
```

File: eng/common/native/init-compiler.sh [hunk 3]
```diff
@@ -84,43 +82,29 @@ if [ -z "$CLR_CC" ]; then
...
-        elif [ "$compiler" = "gcc" ]; then versions="14 13 12 11 10 9 8 7 6 5 4.9"; fi
-
-        for version in $versions; do
-            _major="${version%%.*}"
-            [ -z "${version##*.*}" ] && _minor="${version#*.}"
-            desired_version="$(check_version_exists "$_major" "$_minor")"
-            if [ "$desired_version" != "-1" ]; then majorVersion="$_major"; break; fi
+        maxVersion="$((maxVersion + 1))" # +1 for headspace
+        i="$maxVersion"
+        minVersion=8
```
Comment (am11) [inline eng/common/native/init-compiler.sh:87]:
```suggestion
        minVersion=8
        maxVersion="$((maxVersion + 1))" # +1 for headspace
        i="$maxVersion"
```

**Comment by janvorli** [comment]
I am also worried about a new version automatically. I don't see a big problem in the case when the build fails. The problematic case is when the new compiler has some subtle bug that causes random or rare runtime failures. We have had this kind of problem recently on Windows with MSVC where it was generating instructions that were not supported on some of the intel CPUs that are otherwise supported. Due to those we have decided to stay on a fixed version of the compiler until the new one is fully tested. I'd prefer not getting into this problem on Linux too, especially not with servicing branches.

**Comment by am11** [comment]
@janvorli can you please explain how that is even possible with dotnet-buildtools-prereqs-docker flow? If we think of how all the piece are flowing, it is not possible to pick up the unintended version. We are using these images: https://github.com/dotnet/runtime/blob/5fd965d7bf0196fad051a9cc1e8ebe093ed94fa8/docs/workflow/building/coreclr/linux-instructions.md#docker-images all of them are azurelinux 3.0 based (or the old ones with CBL Mariner). Both Azure Linux and CBL mariner images are skipping this branch because we have our "custom" `clang` without suffix. The suffix version is only relavant for distro source-builds and local environments. Therefore, servicing branches won't be affected.

File: eng/common/native/init-compiler.sh [hunk 4]
```diff
@@ -84,43 +82,29 @@ if [ -z "$CLR_CC" ]; then
...
         if [ -z "$majorVersion" ]; then
             if ! command -v "$compiler" > /dev/null; then
-                echo "Error: No usable version of $compiler found."
+                echo "Error: No compatible version of $compiler was found within the range of $minVersion to $maxVersion. Please upgrade your toolchain or specify the compiler explicitly using CLR_CC and CLR_CXX environment variables."
                 exit 1
             fi
 
             CC="$(command -v "$compiler" 2> /dev/null)"
             CXX="$(command -v "$cxxCompiler" 2> /dev/null)"
             set_compiler_version_from_CC
```
Comment (am11) [inline eng/common/native/init-compiler.sh:102]: @janvorli this is the branch which all the official/service pipelines hit., which remained untouched.

**Comment by janvorli** [comment]
@am11 I didn't remember that we are using self-built clang of a specific version in the CI and for the official builds. So what I was afraid of is not a problem.

Then I don't have any worries about the change other than that people building .NET on their own machines can bump into subtle issues with their builds that we would not know about. Long time ago, I was trying to verify each newly added clang version using the coreclr / libraries tests to make sure they pass. But I've stopped doing that around version 12 and we've kept adding new versions possibly without full verification (unless you were doing that, of course). So this change doesn't change much of the risks.

**Comment by am11** [comment]
@janvorli, yup, since https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/832, we moved the official and other CI legs to just use `clang` without version suffix. Now it is so when we update clang in cbl/azurelinux coredeps file, it just gets updated everywhere and we rely on testing (initial version was 12 now azurelinux images have v18). Also, we are now using floating container tags e.g. `azurelinux-3.0-cross-arm64-net9.0` instead of pinned `azurelinux-3.0-cross-arm64-net9.0-20240712200413-56767ed`, which presents much higher chance of subtle breakages than the old approach; pinned clang version and container tags. The current approach is to fix the issues as they are discovered.

Compared to that, the local or source build selecting latest clang (iff multiple clang versions are installed) poses no risk as those user can pass `-clang18` or `-gcc14` etc. to build.sh on the off-chance there is a need to do so.

**Review by janvorli — COMMENTED** [review]
LGTM, thank you!

**Comment by akoeplinger** [comment]
I merged https://github.com/dotnet/arcade/pull/14963, thanks!

**Pull request closed without merging**